src/gui/editorwidgets/qgstexteditwrapper.cpp

```diff
@@ -74,6 +77,20 @@ QVariant QgsTextEditWrapper::value() const
     // input rather then discarding it entirely
     return QVariant( v.left( field().length() ) );
   }
+  else if ( field().type() == QVariant::Map )
+  {
+    json j = json::parse( v.toUtf8(), nullptr, false );
+    if ( j.type() == json::value_t::discarded ) 
+    {
+      return v;
+    }
+    else 
+    {
+      QJsonDocument qjson = QJsonDocument::fromJson( QString::fromStdString( j.dump( 0 ) ).toUtf8() );
```

Comment:
This method should normally return the same type that was set with `setValue` / the type imposed by `field.type()`, since no provider uses QJsonDocument, this one also doesn't seem right.
It should IMO return a QVariantMap from the parsed string here.

Reply:
OK, I am not clear about the wider reasons were for mapping the Postgres **json** data type to QVariantMap instead of QVariant of type QJsonDocument. I had assumed it was what was available at the time, but it looks like QJsonDocument was available when [this change was introduced ](https://github.com/qgis/QGIS/commit/f92a2ad1c0327981d5f4daab6956c06640e65622)but QVariantMap was the chosen QT container. So presumably we still want to use QVariantMap going forward. Does that sound right to you? And tagging @signedav

Reply:
The reasoning behind: it's an implementation detail how providers save unstructured information.

On widget level we want to have homogeneous data as much as possible and we don't care if the data (QVariantMap / QVariantList) is saved in the provider as json, jsonb, hstore, array, some sort of standardized xml, some native new brilliant binary blob format of tomorrow.

Reply:
... or is there an advantage of QJsonDocument over QVariantMap?

Reply:
Only one I can think of and not totally sure it is possible is that a QJsonDocument can store literal values, which are valid in JSON. Not sure that the other two types can. That would be a strange use of JSON though. Anyway, I have changed it to be QVariantMap and QVariantList.

Reply:
A single string instead of an object? That's possible with QVariant, see line https://github.com/qgis/QGIS/pull/30758/files#diff-34f1e5fd8ed18c81f00b385a9dd233efR96 ;P

Reply:
We should try to avoid `QJson*` methods and stick to QVariant for internal handling, when (un)marshalling is required we are moving to `json::` library because it has better support for numeric types and it's faster.

Reply:
@elpaso I may have misunderstood, but the variant (when it comes in in this instance from a text edit) is a QVariant of type QString, so I wasn't aware of any method to turn it into the required QVariantMap apart from making a QJsonDocument and then doing the conversion. The code above is actually obsolete, and QJsonDocument is just used to construct the QVariantMap.

Reply:
Did you consider using `static QVariant parseJson( const QString &jsonString )` ?